vsintegration/src/FSharp.Editor/Refactor/AddExplicitReturnType.fs

```diff
@@ -0,0 +1,129 @@
+
+        let res =
+            (not funcOrValue.IsValue || not isLambdaIfFunction)
+            && not (funcOrValue.ReturnParameter.Type.IsUnresolved)
+            && not (parseFileResults.IsTypeAnnotationGivenAtPosition symbolUse.Range.Start)
+            && not typeAnnotationRange.IsNone
+
+        match res with
+        | true -> Some funcOrValue
+        | false -> None
+
+    static member refactor
+        (context: CodeRefactoringContext)
+        (symbolUse: FSharpSymbolUse, memberFunc: FSharpMemberOrFunctionOrValue, parseFileResults: FSharpParseFileResults)
+        =
+        let title = SR.AddExplicitReturnTypeAnnotation()
+
+        let getChangedText (sourceText: SourceText) =
+            let inferredType = memberFunc.ReturnParameter.Type.TypeDefinition.DisplayName
+
+            let rangeOfReturnType =
+                parseFileResults.TryRangeOfReturnTypeHint(symbolUse.Range.Start, false)
+
+            let textChange =
+                rangeOfReturnType
+                |> Option.map (fun range -> RoslynHelpers.FSharpRangeToTextSpan(sourceText, range))
+                |> Option.map (fun textSpan -> TextChange(textSpan, $":{inferredType}"))
+
+            match textChange with
+            | Some textChange -> sourceText.WithChanges(textChange)
+            | None -> sourceText
+
+        let codeActionFunc: CancellationToken -> Task<Document> =
+            fun (cancellationToken: CancellationToken) ->
+                task {
+                    let! sourceText = context.Document.GetTextAsync(cancellationToken)
+                    let changedText = getChangedText sourceText
+
+                    let newDocument = context.Document.WithText(changedText)
+                    return newDocument
+                }
+
+        let codeAction = CodeAction.Create(title, codeActionFunc, title)
+
+        do context.RegisterRefactoring(codeAction)
+
+    static member ofFSharpMemberOrFunctionOrValue(symbol: FSharpSymbol) =
+        match symbol with
+        | :? FSharpMemberOrFunctionOrValue as v -> Some v
+        | _ -> None
+
+    override _.ComputeRefactoringsAsync context =
+        let ct = context.CancellationToken
+
+        cancellableTask {
+            let document = context.Document
+            let position = context.Span.Start
+            let! sourceText = document.GetTextAsync()
+            let textLine = sourceText.Lines.GetLineFromPosition position
+            let textLinePos = sourceText.Lines.GetLinePosition position
+            let fcsTextLineNumber = Line.fromZ textLinePos.Line
+
+            let! lexerSymbol =
+                document.TryFindFSharpLexerSymbolAsync(position, SymbolLookupKind.Greedy, false, false, nameof (AddExplicitReturnType))
+
+            let! (parseFileResults, checkFileResults) = document.GetFSharpParseAndCheckResultsAsync(nameof (AddExplicitReturnType))
+
+            let res =
+                internalOption {
+                    let! lexerSymbol = lexerSymbol
```

Comment:
@psfinaki I still prefer the internalOption after returning it to non option:
```f#
let symbolUseOpt =
    lexerSymbol
    |> Option.bind (fun lexer ->
        checkFileResults.GetSymbolUseAtLocation(
            fcsTextLineNumber,
            lexer.Ident.idRange.EndColumn,
            textLine.ToString(),
            lexer.FullIsland
        ))

let memberFuncOpt =
    symbolUseOpt
    |> Option.bind (fun sym -> sym.Symbol |> AddExplicitReturnType.ofFSharpMemberOrFunctionOrValue)

match (symbolUseOpt, memberFuncOpt) with
| (Some symbolUse, Some memberFunc) ->
    let isValidMethod =
        memberFunc
        |> AddExplicitReturnType.isValidMethodWithoutTypeAnnotation symbolUse parseFileResults

    match isValidMethod with
    | Some memberFunc -> do AddExplicitReturnType.refactor context (symbolUse, memberFunc, parseFileResults)
    | None -> ()
| _ -> ()

return ()
``` 

But maybe this isnt the best way of writing it or I am missing something?

Reply:
Ok. Damn |>! XD I concede that point. It is reverted